### DISCUSSION TITLE
How to Optionally use Dependency injection

### DISCUSSION THREAD

**Discussion by shyamsukhamit** [fastapi/fastapi#6610] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Any, List, Optional
from fastapi import APIRouter, Body, Depends, HTTPException, File, UploadFile, Form, Security
from sqlalchemy.orm import Session
from fastapi.encoders import jsonable_encoder
from app import crud, models, schemas
from app.api import deps
from app.core.config import settings
from app.db.database import get_db
from datetime import datetime, date, time, timedelta
import shutil
from uuid import UUID
import time
import os, pathlib
from app.api.utils import aws_upload
import uuid
import secrets
router = APIRouter()

@router.get("/cart/{tmp_id}", response_model=Any)
def cart(db: Session = Depends(get_db),  user_id:Optional[UUID] = Depends(deps.get_current_active_user),
) -> models.Cart:

	return crud.crud_Cart.get_self(db, user_id=user_id, limit=settings.USER_PER_PAGE, page=page)
```

### Description

My use case is how to optionally call Dependency injection of current_user - as I want to fetch the cart items via 

1. tmp_id when user is not logged in
2. user_id when user is logged in

I am trying to call the Depends(deps.get_current_active_user) optionally but of no use.

Can anyone help me in finding the correct way to implement this logic or any other way to try to get the problem resolved

### Operating System

Windows

### Operating System Details

Windows 10

### FastAPI Version

0.68.1

### Python Version

3.8.6

**Comment by STeveShary** [comment, marked as answer]
I think if you had a separate dependency, you can use what you need:
```python3

def get_temporary_user_cart_id()
	pass

def get_cart_user_id(active_user:Optional[UUID] = Depends(deps.get_current_active_user)) -> UUID:
	if not active_user:
		return get_temporary_user_cart_id() 
	return active_user


@router.get("/cart/{tmp_id}", response_model=Any)
def cart(db: Session = Depends(get_db),  user_id:Optional[UUID] = Depends(get_cart_user_id),
) -> models.Cart:

	return crud.crud_Cart.get_self(db, user_id=user_id, limit=settings.USER_PER_PAGE, page=page)
```

Does this help?

**Reply by YuriiMotov** [reply]
And you can use `auto_error=False` to avoid having auth error if user is not authenticated

**Comment by shyamsukhamit** [comment]
@STeveShary   This will not help, as at the end OAuth2PasswordBearer for getting and verifying the token give unauthenticated error,
and I am unable to find any other way to get the token while the user is logged in.

Getting the tmp_id is not an issue as I will get via query/path params, but when user is logged in and Depends(get_current_active_user) is called which in turns calls Depends(get_current_user), which in turn calls Depends(OAuth2PasswordBearer) and at the end it raises and exception .


I could not find any ways to just check if user is logged in, without accessing the token.

Any help is this regard would be quite a help.

**Comment by shyamsukhamit** [comment]
I solved it by using request and getting the token from request headers and then validation via jwt.decode
